### PR TITLE
Semantic segmentation class weights

### PR DESCRIPTION
**Pull request by Y-T-G**
Semantic segmentation does not support class weighting, but there are hardcoded weights when training on Cityscapes.

This PR adds class weighting support through `cls_pw`. Hardcoded weights are still used when `cls_pw` isn't specified or is 0, so default training is unaffected.

Weights are calculated from ENet formula 1/ln(1.02 + frequency) (Paszke et al., 2016). Different from inverse class frequency formula used for other tasks. Used this because it seems more appropriate for semantic segmentation.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
This PR improves how Ultralytics computes and applies class weights during training, especially for imbalanced datasets, with new support for semantic segmentation and safer handling of missing label data ⚖️🧠

### 📊 Key Changes
- Added reusable helper methods in detection training to:
  - count per-class labels from the training set
  - compute inverse-frequency class weights automatically
- Updated detection training to safely skip class-weighting when no valid class counts are found, avoiding failures with unreadable or missing labels 🛡️
- Added semantic segmentation class-weight support for multi-class datasets by counting pixel frequencies from segmentation masks 🎨
- Introduced segmentation-specific class-weight calculation using a pixel-frequency-based weighting strategy, better suited to semantic masks than detection-style counting
- Reused the new semantic mask counting logic for label plotting, reducing duplicate code and improving consistency
- Updated detection loss so classification BCE loss can be multiplied by learned class weights during training 📉
- Updated semantic segmentation loss to use:
  - model-provided class weights when available
  - existing Cityscapes preset weights as a fallback
  - no class weighting for binary segmentation, by design
- Bumped the package version from `8.4.67` to `8.4.68` 🚀

### 🎯 Purpose & Impact
- Helps models learn better on imbalanced datasets, where some classes appear much less often than others ⚖️
- Improves semantic segmentation training by using pixel-level statistics instead of object-level counts, which is more appropriate for mask-based tasks
- Makes training more robust when dataset labels or masks are missing, invalid, or unreadable ✅
- Brings more consistent class-weight behavior across detection and segmentation tasks
- Can improve performance on rare classes, though training results may shift for users who enable `cls_pw` and train on uneven datasets 📈

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Y-T-G, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help with review readiness—an Ultralytics engineer will assist you soon. Please review the checklist below to help ensure a smooth integration. 😊

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The refactor is clean overall and the shared helper approach makes sense, but there are two functional gaps to address: zero/invalid semantic class counts currently suppress the intended Cityscapes fallback, and binary semantic segmentation still ignores the computed class weights.

💬 Posted 2 inline comments

File: ultralytics/utils/loss.py [hunk 1]
```diff
@@ -1289,13 +1289,16 @@ def __init__(self, model: torch.nn.Module):
         self.dtype = next(model.parameters()).dtype
         data_name = Path(str(getattr(model.args, "data", "") or "")).stem.lower()
         self.use_cityscapes_weight = data_name in {"cityscapes", "cityscapes8"} and self.nc == len(CITYSCAPES_WEIGHT)
+        weight = getattr(model, "class_weights", None)  # cls_pw frequency weights, else hardcoded Cityscapes
```
Comment (Y-T-G) [inline ultralytics/utils/loss.py:1295]: Intentional

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 2

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The refactor is clean and the Cityscapes fallback behavior looks preserved for multi-class segmentation, but binary semantic segmentation is still not supported end-to-end: the new histogram code counts the wrong pixels when `nc == 1`, and the loss setup never applies the computed weights in the BCE path.

💬 Posted 2 inline comments

File: ultralytics/models/yolo/semantic/train.py [hunk 2]
```diff
@@ -74,48 +74,52 @@ def get_validator(self):
...
         mask_files = getattr(dataset, "mask_files", [])
         if not mask_files:
-            LOGGER.warning("No semantic mask files found, skipping label plot.")
-            return
-
-        sample_size = min(1000, len(mask_files))
-        indices = np.linspace(0, len(mask_files) - 1, sample_size).astype(int)
-
+            return pixel_counts
+        indices = np.linspace(0, len(mask_files) - 1, min(1000, len(mask_files))).astype(int)
```
Comment (UltralyticsAssistant) [inline ultralytics/models/yolo/semantic/train.py]: 💡 **MEDIUM**: This helper is now used for training-time class weights as well as plotting, but it still caps the scan at 1000 masks. On large datasets that can materially misestimate frequencies, and any class missed by the sample gets `p=0` in `compute_class_weights()`, which assigns it the maximum ENet weight. That can noticeably distort the CE term. I'd keep the 1000-mask cap for `plot_training_labels()`, but compute weights from the full training set here (or via a separate full-scan helper).

File: ultralytics/utils/loss.py [hunk 3]
```diff
@@ -1289,13 +1289,16 @@ def __init__(self, model: torch.nn.Module):
...
+        if weight is None and self.use_cityscapes_weight:
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/loss.py:1296]: 💡 **MEDIUM**: With this fallback, Cityscapes weights still apply when a user explicitly sets `cls_pw=0`, because `set_class_weights()` returns early and leaves `model.class_weights` unset. That makes it impossible to opt out of weighting on Cityscapes and doesn't match the PR behavior described in the summary (`hardcoded weights` only when `cls_pw` isn't specified). Consider distinguishing the default/unset case from an explicit zero, or setting an explicit unity weight tensor when weighting is disabled.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 3

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The refactor is clean overall and the shared helper approach makes the weighting logic easier to reuse. I found two logic issues worth addressing before merge: the 1000-mask sampling cap is too lossy for training-time weight computation, and the new fallback path still forces Cityscapes weights when `cls_pw=0`, so users cannot explicitly disable weighting.

💬 Posted 2 inline comments

File: ultralytics/models/yolo/semantic/train.py (hunk 2 above)
Comment (Y-T-G) [reply ultralytics/models/yolo/semantic/train.py]: Intentional

**Review by glenn-jocher — APPROVED** [review]
Reviewed. The shared weighting path is compact, and the follow-up keeps training-time class weights based on full semantic mask counts while preserving sampled label plots.

**Comment by UltralyticsAssistant** [comment]
Merged! 🎉 Thanks @Y-T-G for this thoughtful improvement, with support from @glenn-jocher.

“Details make perfection, and perfection is not a detail.” — Leonardo da Vinci

That fits this PR well: better class-weight handling, safer behavior with missing labels, and segmentation-aware weighting all strengthen training where it matters most, especially on imbalanced datasets. These kinds of careful improvements make Ultralytics more robust, more consistent, and more effective for real-world use.

Really appreciate the craftsmanship here — thank you both for helping move the project forward.
